scipy/stats/tests/test_hypotests.py

```diff
@@ -405,3 +406,105 @@ def test_contingency_table_return(self):
+    @pytest.mark.parametrize("input_sample,expected", [
+        ([[43, 40], [10, 39]], (-3.920362887717, 0.000289470662)),
+        ([[100, 2], [1000, 5]], (1.139432816087, 0.950272080594)),
+        ([[2, 7], [8, 2]], (3.079373904042, 0.020172119141)),
+        ([[5, 1], [10, 10]], (-1.622375939458, 0.150599922226)),
+        ([[5, 15], [20, 20]], (1.974771239528, 0.063038448651)),
+        ([[5, 16], [20, 25]], (1.722122973346, 0.133329494287)),
+        ([[10, 5], [10, 1]], (1.765469659009, 0.250566655215)),
+        ([[5, 0], [1, 4]], (-5.477225575052, 0.007812500000)),
+        ([[0, 1], [3, 2]], (1.224744871392, 0.509667991877)),
+        ([[0, 2], [6, 4]], (1.732050807569, 0.197019618792)),
+        ([[2, 7], [8, 2]], (3.079373904042, 0.020172119141)),
+    ])
+    def test_pooled_param(self, input_sample, expected):
+        """The expected values have been generated by R, using a resolution
+        for the nuisance parameter of 1e-6.
+        """
+        res = barnard_exact(input_sample, pooled=False)
+        TX_obs, pvalue = res.statistic, res.pvalue
+        assert_allclose([TX_obs, pvalue], expected)
+
+    def test_raises(self):
+        # test we raise an error for wrong input number of nuisances.
+        error_msg = (
+            "Number of iterations `num_it` must be strictly positive,"
+            " found 0")
+        with assert_raises(ValueError, match=error_msg):
+            barnard_exact([[1, 2], [3, 4]], num_it=0)
+
+        # test we raise an error for wrong shape of input.
+        error_msg = "The input `table` must be of shape \\(2, 2\\)."
+        with assert_raises(ValueError, match=error_msg):
+            barnard_exact(np.arange(6).reshape(2, 3))
+
+        # Test all values must be positives
+        error_msg = "All values in `table` must be nonnegative."
+        with assert_raises(ValueError, match=error_msg):
+            barnard_exact([[-1, 2], [3, 4]])
+
+        # Test value error on wrong alternative param
+        error_msg = (
+            "`alternative` should be one of {'two-sided', 'less', 'greater'},"
+            " found .*")
+        with assert_raises(ValueError, match=error_msg):
+            barnard_exact([[1, 2], [3, 4]], "not-correct")
+
+    @pytest.mark.parametrize("input_sample,expected", [
+        ([[0, 5], [0, 10]], (1.0, np.nan)),
+        ([[5, 0], [10, 0]], (1.0, np.nan)),
+    ])
+    def test_row_or_col_zero(self, input_sample, expected):
+        res = barnard_exact(input_sample)
+        TX_obs, pvalue = res.statistic, res.pvalue
+        assert_equal(pvalue, expected[0])
+        assert_equal(TX_obs, expected[1])
+
+    @pytest.mark.slow
+    @pytest.mark.parametrize("input_sample,expected", [
+        ([[2, 7], [8, 2]], (2.518474945157, 0.009886140845,
+        1 - 0.009886140845)),
+        ([[200, 7], [8, 300]], (-21.320036698460, 0.0, 1.0)),
+        ([[28, 21], [6, 1957]], (-30.489638143953, 0.0, 1.0)),
+    ])
+    def test_less_greater(self, input_sample, expected):
```

Comment:
Where do these ones come from?

Reply:
It comes from R, but I am working on changing it anyway :-)